### PR TITLE
Cleanup Netty bootstrap options.

### PR DESCRIPTION
**Pull request by johnou**

### CODE REVIEW

File: messaging/src/main/java/io/atomix/messaging/impl/NettyMessagingService.java [hunk 1]
```diff
@@ -525,14 +525,15 @@ private Bootstrap bootstrapClient(Endpoint endpoint) {
     Bootstrap bootstrap = new Bootstrap();
     bootstrap.option(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT);
     bootstrap.option(ChannelOption.WRITE_BUFFER_WATER_MARK,
-        new WriteBufferWaterMark(10 * 32 * 1024, 10 * 64 * 1024));
-    bootstrap.option(ChannelOption.SO_SNDBUF, 1048576);
+            new WriteBufferWaterMark(10 * 32 * 1024, 10 * 64 * 1024));
+    bootstrap.option(ChannelOption.SO_RCVBUF, 1024 * 1024);
+    bootstrap.option(ChannelOption.SO_SNDBUF, 1024 * 1024);
+    bootstrap.option(ChannelOption.SO_KEEPALIVE, true);
```
Comment (johnou) [inline messaging/src/main/java/io/atomix/messaging/impl/NettyMessagingService.java]: functional change

File: messaging/src/main/java/io/atomix/messaging/impl/NettyMessagingService.java [hunk 2]
```diff
@@ -525,14 +525,15 @@ private Bootstrap bootstrapClient(Endpoint endpoint) {
     Bootstrap bootstrap = new Bootstrap();
     bootstrap.option(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT);
     bootstrap.option(ChannelOption.WRITE_BUFFER_WATER_MARK,
-        new WriteBufferWaterMark(10 * 32 * 1024, 10 * 64 * 1024));
-    bootstrap.option(ChannelOption.SO_SNDBUF, 1048576);
+            new WriteBufferWaterMark(10 * 32 * 1024, 10 * 64 * 1024));
+    bootstrap.option(ChannelOption.SO_RCVBUF, 1024 * 1024);
```
Comment (johnou) [inline messaging/src/main/java/io/atomix/messaging/impl/NettyMessagingService.java]: was SO_RCVBUF omitted intentionally?

File: messaging/src/main/java/io/atomix/messaging/impl/NettyMessagingService.java [hunk 3]
```diff
@@ -545,9 +546,12 @@ private Bootstrap bootstrapClient(Endpoint endpoint) {
   private CompletableFuture<Void> startAcceptingConnections() {
     CompletableFuture<Void> future = new CompletableFuture<>();
     ServerBootstrap b = new ServerBootstrap();
+    b.option(ChannelOption.SO_REUSEADDR, true);
```
Comment (johnou) [inline messaging/src/main/java/io/atomix/messaging/impl/NettyMessagingService.java]: functional change

File: messaging/src/main/java/io/atomix/messaging/impl/NettyMessagingService.java [hunk 4]
```diff
@@ -545,9 +546,12 @@ private Bootstrap bootstrapClient(Endpoint endpoint) {
...
+    b.option(ChannelOption.SO_BACKLOG, 128);
     b.childOption(ChannelOption.WRITE_BUFFER_WATER_MARK,
-        new WriteBufferWaterMark(8 * 1024, 32 * 1024));
-    b.option(ChannelOption.SO_RCVBUF, 1048576);
```
Comment (johnou) [inline messaging/src/main/java/io/atomix/messaging/impl/NettyMessagingService.java]: I believe SO_RCVBUF should be set on the child option, not option for server bootstrap.

**Comment by johnou** [comment]
@normanmaurer any other obvious / non-obvious options you can recommend? I'm having a hard time matching JGroups performance with Netty.

**Comment by kuujo** [comment]
@johnou maybe take a look around in the Vert.x event bus. I’m sure a lot of @normanmaurer’s old Netty code is still hanging around there behind a very similar abstraction.

When in doubt, look at Norman’s Netty code :-P

File: messaging/src/main/java/io/atomix/messaging/impl/NettyMessagingService.java [hunk 5]
```diff
@@ -525,14 +525,16 @@ private Bootstrap bootstrapClient(Endpoint endpoint) {
     Bootstrap bootstrap = new Bootstrap();
     bootstrap.option(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT);
     bootstrap.option(ChannelOption.WRITE_BUFFER_WATER_MARK,
-        new WriteBufferWaterMark(10 * 32 * 1024, 10 * 64 * 1024));
-    bootstrap.option(ChannelOption.SO_SNDBUF, 1048576);
+            new WriteBufferWaterMark(10 * 32 * 1024, 10 * 64 * 1024));
+    bootstrap.option(ChannelOption.SO_RCVBUF, 1024 * 1024);
+    bootstrap.option(ChannelOption.SO_SNDBUF, 1024 * 1024);
+    bootstrap.option(ChannelOption.SO_KEEPALIVE, true);
+    bootstrap.option(ChannelOption.TCP_NODELAY, true);
```
Comment (johnou) [inline messaging/src/main/java/io/atomix/messaging/impl/NettyMessagingService.java]: Nagles must be disabled on client and server for it to work.

**Comment by johnou** [comment]
@kuujo these changes should be good to merge.

File: messaging/src/main/java/io/atomix/messaging/impl/NettyMessagingService.java (hunk 5 above)
Comment (kuujo) [reply messaging/src/main/java/io/atomix/messaging/impl/NettyMessagingService.java]: Nice

**Pull request merged**